### PR TITLE
feat: Ability to Hyperlink text in Checkbox on Additional Questions

### LINKED ISSUE #14163: [CAL-3292] Ability to Hyperlink text in Checkbox on Additional Questions
Many times, users need a checkbox for their attendees to agree to terms and conditions. They want to add URLs that are hyperlinked, so that attendees can read and agree to these conditions. 

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-3292](https://linear.app/calcom/issue/CAL-3292/ability-to-hyperlink-text-in-checkbox-on-additional-questions)</sub>

### PR DESCRIPTION
**Pull request by anikdhabal**
## What does this PR do?

- Fixes #14163 

Screenshot 2024-05-25 021856

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

**Review by Amit91848 — DISMISSED** [review]
Great work so far Anik. Just a small issue. Labels aren't being rendered properly when it has links.

Screenshot 2024-06-27 192042

Screenshot 2024-06-27 192154

File: packages/features/form-builder/FormBuilder.tsx [hunk 1]
```diff
@@ -483,16 +488,36 @@ function FieldEditDialog({
...
-                    />
+                    <div>
+                      {fieldForm.getValues("type") === "boolean" ? (
+                        <div className="mt-6">
+                          <Label>{t("label")}</Label>
+                          <Editor
+                            getText={() => md.render(fieldForm.getValues("label") || "")}
+                            setText={(value: string) => {
+                              fieldForm.setValue("label", turndown(value), { shouldDirty: true });
+                              console.log(turndown(value));
```
Comment (Amit91848) [inline packages/features/form-builder/FormBuilder.tsx:499]: NIT: Remove log

**Comment by anikdhabal** [comment]
Thanks for the review @Amit91848. I will fix that.

File: packages/features/form-builder/FormBuilder.tsx [hunk 2]
```diff
@@ -483,16 +488,36 @@ function FieldEditDialog({
...
-                      label={t("label")}
-                    />
+                    <div>
+                      {fieldForm.getValues("type") === "boolean" ? (
+                        <div className="mt-6">
+                          <Label>{t("label")}</Label>
+                          <Editor
+                            getText={() => md.render(fieldForm.getValues("label") || "")}
+                            setText={(value: string) => {
+                              fieldForm.setValue("label", turndown(value), { shouldDirty: true });
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:498]: Can we abstract this component out in a CheckboxField or something similar component

File: packages/features/form-builder/FormBuilder.tsx [hunk 3]
```diff
@@ -438,9 +441,11 @@ function FieldEditDialog({
   const variantsConfig = fieldForm.watch("variantsConfig");
 
   const fieldTypes = Object.values(fieldTypesConfigMap);
+  const [firstRender, setFirstRender] = useState(true);
 
   return (
-    <Dialog open={dialog.isOpen} onOpenChange={onOpenChange}>
+    <Dialog open={dialog.isOpen} onOpenChange={onOpenChange} modal={false}>
+      <div className="fadeIn fixed inset-0 z-50 bg-neutral-800 bg-opacity-70 transition-opacity dark:bg-opacity-70" />
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:448]: Why is this changed?

File: packages/features/form-builder/FormBuilder.tsx [hunk 4]
```diff
@@ -483,16 +488,36 @@ function FieldEditDialog({
...
-                      // System fields have a defaultLabel, so there a label is not required
-                      required={
-                        !["system", "system-but-optional"].includes(fieldForm.getValues("editable") || "")
-                      }
-                      placeholder={t(fieldForm.getValues("defaultLabel") || "")}
-                      containerClassName="mt-6"
-                      label={t("label")}
-                    />
+                    <div>
+                      {fieldForm.getValues("type") === "boolean" ? (
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:492]: Don't we have `fieldForm.getValues("type")` saved in some variable?

**Review by hariombalhara — CHANGES_REQUESTED** [review]
@anikdhabal Left a few suggestions

File: packages/features/form-builder/FormBuilder.tsx (hunk 3 above)
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx:448]: Hey, when modal={true} is set to true (default), interaction with outside elements will be disabled, and only the dialog content will be visible to screen readers. For that reason, we can't make text a link in the editor, as it opens another tooltip. So, I set modal to false and externally add the dialog styles that are present by default, because these styles are removed when modal is set to false.

File: packages/features/form-builder/FormBuilder.tsx (hunk 4 above)
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx]: Not currently, but I’ve added it as it is used many times

File: packages/features/form-builder/FormBuilder.tsx (hunk 2 above)
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx]: Done

File: packages/features/form-builder/FormBuilder.tsx [hunk 5]
```diff
@@ -423,28 +426,53 @@ function FieldEditDialog({
...
+
+  const CheckboxFieldLabel = () => {
+    return (
+      <div className="mt-6">
+        <Label>{t("label")}</Label>
+        <Editor
+          getText={() => md.render(fieldForm.getValues("label") || "")}
+          setText={(value: string) => {
+            fieldForm.setValue("label", turndown(value), { shouldDirty: true });
+            console.log(turndown(value));
```
Comment (hariombalhara) [inline packages/features/form-builder/FormBuilder.tsx:461]:
```suggestion
```

File: packages/features/form-builder/FormBuilder.tsx (hunk 3 above)
Comment (hariombalhara) [reply packages/features/form-builder/FormBuilder.tsx:475]: This seems hackish. Let's figure out a better way to handle it?

Hackish because we are using the internal details of dialog to make it look like it.

File: packages/features/form-builder/FormBuilder.tsx (hunk 5 above)
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx]: Done

File: packages/features/form-builder/FormBuilder.tsx (hunk 3 above)
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx]: Yeah, update it

File: apps/web/modules/bookings/views/bookings-single-view.tsx [hunk 6]
```diff
@@ -645,7 +646,12 @@ export default function Success(props: PageProps) {
 
                           return (
                             <>
-                              <div className="text-emphasis mt-4 font-medium">{label}</div>
+                              <div
+                                className="text-emphasis mt-4 font-medium"
+                                dangerouslySetInnerHTML={{
+                                  __html: markdownToSafeHTML(label),
```
Comment (hariombalhara) [inline apps/web/modules/bookings/views/bookings-single-view.tsx:652]: `markdownToSafeHTML` has huge bundle size.
image

Can we run it server side only

Comment (anikdhabal) [reply apps/web/modules/bookings/views/bookings-single-view.tsx:652]: Yeah, this make sense👍

**Review by hariombalhara — CHANGES_REQUESTED** [review]
https://www.loom.com/share/4d43f5540fb04d4a94d4e63833f6faae

The text is getting duplicated when trying to edit the label.

File: packages/features/form-builder/FormBuilder.tsx [hunk 7]
```diff
@@ -428,29 +431,52 @@ function FieldEditDialog({
...
+          }}
+          excludedToolbarItems={["blockType", "bold", "italic"]}
+          disableLists
+          firstRender={firstRender}
+          setFirstRender={setFirstRender}
+          placeholder={t(fieldForm.getValues("defaultLabel") || "")}
+        />
+      </div>
+    );
+  };
```
[the remark this answers is not in the document]
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx:475]: Thanks for explaining @hariombalhara , but di we really need make this a component? This has only one use case.

[the remark this answers is not in the document]
Comment (hariombalhara) [reply packages/features/form-builder/FormBuilder.tsx:475]: It is already a component as it is used in JSX as a tag.

[the remark this answers is not in the document]
Comment (hariombalhara) [reply packages/features/form-builder/FormBuilder.tsx:475]: What I meant was moving it outside of existing component.SO, it can still stay in this module.

[the remark this answers is not in the document]
Comment (anikdhabal) [reply packages/features/form-builder/FormBuilder.tsx:475]: > It is already a component as it is used in JSX as a tag.

Yeah, my bad. I am wanted to ask that do we really need to abstact out this component? There no use cases of it and need to ass and import planty of things.

File: packages/ui/components/dialog/Dialog.tsx [hunk 8]
```diff
@@ -109,6 +113,9 @@ export const DialogContent = React.forwardRef<HTMLDivElement, DialogContentProps
     return (
       <Portal>
         <Overlay className="fadeIn fixed inset-0 z-50 bg-neutral-800 bg-opacity-70 transition-opacity dark:bg-opacity-70 " />
+        {modalDisabled && (
```
Comment (hariombalhara) [inline packages/ui/components/dialog/Dialog.tsx:116]: How about we call it `forceOverlayWhenNoModal` to make it clear its use.

i would have preferred to be able to use Dialog.modal prop here using Dialog context but that seems to be not provided by radix-ui.

Comment (anikdhabal) [reply packages/ui/components/dialog/Dialog.tsx:116]: Okay 👍, nah it's not provided by them.

**Comment by anikdhabal** [comment]
> The text is getting duplicated when trying to edit the label.

Can't reproduce

https://github.com/user-attachments/assets/b20f9396-b211-4233-8f8c-686ff8f40f18

File: packages/features/shell/Shell.tsx [hunk 9]
```diff
@@ -215,7 +215,7 @@ const useBanners = () => {
 const Layout = (props: LayoutProps) => {
   const banners = useBanners();
   const pathname = usePathname();
-  const isFullPageWithoutSidebar = pathname.startsWith("/apps/routing-forms/reporting/");
+  const isFullPageWithoutSidebar = pathname?.startsWith("/apps/routing-forms/reporting/");
```
Comment (anikdhabal) [inline packages/features/shell/Shell.tsx:218]: This is not part of this pr, but this type error is the reason currently user faced type error during build

**Review by hariombalhara — DISMISSED** [review]
LGTM now !!

**Comment by hariombalhara** [comment]
Form builder related e2e tests seem to be failng @anikdhabal

File: apps/web/playwright/manage-booking-questions.e2e.ts [hunk 10]
```diff
@@ -72,9 +72,9 @@ test.describe("Manage Booking Questions", () => {
         await addQuestionAndSave({
           page,
           question: {
-            name: "agree-to-terms",
-            type: "Checkbox",
-            label: "Agree to [terms](https://example.com/terms)",
```
Comment (hariombalhara) [inline apps/web/playwright/manage-booking-questions.e2e.ts:77]: Why are we removing this test?

File: apps/web/pages/auth/login.tsx [hunk 11]
```diff
@@ -256,12 +256,10 @@ inferSSRProps<typeof getServerSideProps> & WithNonceProps<{}>) {
                     CustomStartIcon={<GoogleIcon />}
                     onClick={async (e) => {
                       e.preventDefault();
-                      const res = await signIn("google", {
+                      setLastUsed("google");
+                      await signIn("google", {
```
Comment (Amit91848) [inline apps/web/pages/auth/login.tsx:260]: Are these changes accidental?  If not we should handle it in a separate PR

File: packages/features/auth/SAMLLogin.tsx [hunk 12]
```diff
@@ -26,10 +26,8 @@ export function SAMLLogin({ samlTenantID, samlProductID, setErrorMessage }: Prop
 
   const mutation = trpc.viewer.public.samlTenantProduct.useMutation({
     onSuccess: async (data) => {
-      const res = await signIn("saml", {}, { tenant: data.tenant, product: data.product });
-      if (res && !res.error) {
-        setLastUsed("saml");
-      }
+      setLastUsed("saml");
+      await signIn("saml", {}, { tenant: data.tenant, product: data.product });
```
Comment (Amit91848) [inline packages/features/auth/SAMLLogin.tsx:30]: Same here?

**Review by Amit91848 — COMMENTED** [review]
Great work @anikdhabal. Changes look good otherwise 🔥

File: apps/web/pages/auth/login.tsx (hunk 11 above)
Comment (anikdhabal) [reply apps/web/pages/auth/login.tsx:260]: Not accidental, small improvement for my merged pr. No needed to create a separate pr for this small changes

File: apps/web/playwright/manage-booking-questions.e2e.ts (hunk 10 above)
Comment (anikdhabal) [reply apps/web/playwright/manage-booking-questions.e2e.ts]: Yeah not needed, fixed it

File: apps/web/pages/auth/login.tsx (hunk 11 above)
Comment (hariombalhara) [reply apps/web/pages/auth/login.tsx]: @anikdhabal This isn't a good practice to have unrelated changes in the PR. Two major reasons
1. Unrelated changes might delay the original fix(if reviewer have concern with the unrelated changes code quality/stability)
2. More importantly, the PR can be reverted and that reverts the unrelated change in the PR which people won't be aware of.


So, could you please remove these changes from this PR.

Comment (anikdhabal) [reply apps/web/pages/auth/login.tsx]: Done

**Review by hariombalhara — APPROVED** [review]
